File: binding/development/tools/cast/src/c++/cast/architecture/SubarchitectureWorkingMemoryProtocol.cpp
/*
 * CAST - The CoSy Architecture Schema Toolkit
 *
 * Copyright (C) 2006-2007 Nick Hawes
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA
 *
 */

#include "SubarchitectureWorkingMemoryProtocol.hpp"
#include <sstream>
#include <cstdlib>

using namespace std;

namespace cast {

  string SubarchitectureWorkingMemoryProtocol::ID_PREFIX = cdl::ID_QUERY_PREFIX;
  string SubarchitectureWorkingMemoryProtocol::EXISTS_PREFIX = cdl::EXISTS_QUERY_PREFIX;
  string SubarchitectureWorkingMemoryProtocol::OVERWRITE_COUNT_PREFIX = cdl::OVERWRITE_COUNT_QUERY_PREFIX;
  string SubarchitectureWorkingMemoryProtocol::ID_ARRAY_PREFIX = cdl::ID_ARRAY_QUERY_PREFIX;
  string SubarchitectureWorkingMemoryProtocol::TYPE_PREFIX = cdl::TYPE_QUERY_PREFIX;


  string SubarchitectureWorkingMemoryProtocol::createIDQuery(const string &_srcID, 
							     const string &_srcSA, 
							     const string &_subarch, 
							     const string &_id) {
    return ID_PREFIX + "\\" + _srcID + "\\" + _srcSA + "\\" + _subarch + "\\" + _id;
  }

  string SubarchitectureWorkingMemoryProtocol::createExistsQuery(const string &_srcID, 
								 const string &_srcSA, 
								 const string &_subarch, 
								 const string &_id) {
    return EXISTS_PREFIX + "\\" + _srcID + "\\" + _srcSA + "\\" + _subarch + "\\" + _id;
  }

  string SubarchitectureWorkingMemoryProtocol::createOverwriteCountQuery(const string &_srcID, 
									 const string &_srcSA, 
									 const string &_subarch, 
									 const string &_id) {
    return OVERWRITE_COUNT_PREFIX + "\\" + _srcID + "\\" + _srcSA + "\\" + _subarch + "\\" + _id;
  }


  string SubarchitectureWorkingMemoryProtocol::createTypeQuery(const string &_srcID, 
							       const string &_srcSA, 
							       const string &_subarch, 
							       const string &_type, 
							       const int &_count) {
    std::ostringstream o;
    o << TYPE_PREFIX;
    o << "\\";
    o << _srcID;
    o << "\\";
    o << _srcSA;  
    o << "\\";
    o << _subarch;
    o << "\\";
    o << _count;
    o << "\\";
    o << _type;
    return o.str();

  }

  string SubarchitectureWorkingMemoryProtocol::createIDArrayQuery(const string &_srcID, 
								  const string &_srcSA, 
								  const string &_subarch, 
								  const vector<string> &_ids) {

    string s = ID_ARRAY_PREFIX + "\\" + _srcID + "\\" + _srcSA + "\\" + _subarch + "\\";
    for (unsigned int i = 0; i < _ids.size(); i++) {
      s += _ids[i] + "\\";
    }
    return s;
  
  }

  AbstractWorkingMemoryPullQuery * 
  SubarchitectureWorkingMemoryProtocol::parseQuery(const string &_query) {

    string prefix = _query.substr(0,2);

    //cout<<"prefix: "<<prefix<<endl;

    if (prefix == ID_PREFIX) {
      return extractID(_query);
    }
    else if (prefix == EXISTS_PREFIX) {
      return extractExists(_query);
    }
    else if (prefix == OVERWRITE_COUNT_PREFIX) {
      return extractOverwriteCount(_query);
    }
    else if (prefix == ID_ARRAY_PREFIX) {
      return extractIDArray(_query);
    }
    else if (prefix == TYPE_PREFIX) {
      return extractType(_query);
    }
    else {
      return extractUnknown(_query);
    }
  
  }

  AbstractWorkingMemoryPullQuery * 
  SubarchitectureWorkingMemoryProtocol::extractUnknown(const string &_query) {

    vector<string> tokens;
  
    tokenizeString(_query, tokens, "\\");

    if(tokens.size() < 4) {
      throw(runtime_error("Unknown memory protocol for query: " + _query));
    }

    return new WorkingMemoryPullQuery<string>(tokens[1], tokens[2], tokens[3], 
					      0,
					      new string(_query), 
					      AbstractWorkingMemoryPullQuery::OTHER);
  }

  WorkingMemoryPullQuery<string> * 
  SubarchitectureWorkingMemoryProtocol::extractID(const string &_query) {

    vector<string> tokens;
  
    tokenizeString(_query, tokens, "\\");

    if(5 != tokens.size()) {
      cerr<<"SubarchitectureWorkingMemoryProtocol::extractID incorrect number of tokens in query"<<endl;
      return NULL;
    }

    return new WorkingMemoryPullQuery<string>(tokens[1], tokens[2], tokens[3], 
					      0,
					      new string(tokens[4]), 
					      AbstractWorkingMemoryPullQuery::ID);
  }

  WorkingMemoryPullQuery<string> * 
  SubarchitectureWorkingMemoryProtocol::extractExists(const string &_query) {

    vector<string> tokens;
  
    tokenizeString(_query, tokens, "\\");

    if(5 != tokens.size()) {
      cerr<<"SubarchitectureWorkingMemoryProtocol::extractExists incorrect number of tokens in query"<<endl;
      return NULL;
    }

    return new WorkingMemoryPullQuery<string>(tokens[1], tokens[2], tokens[3], 
					      0,
					      new string(tokens[4]), 
					      AbstractWorkingMemoryPullQuery::EXISTS);
  }

  WorkingMemoryPullQuery<string> * 
  SubarchitectureWorkingMemoryProtocol::extractOverwriteCount(const string &_query) {

    vector<string> tokens;
  
    tokenizeString(_query, tokens, "\\");

    if(5 != tokens.size()) {
      cerr<<"SubarchitectureWorkingMemoryProtocol::extractOverwriteCount incorrect number of tokens in query"<<endl;
      return NULL;
    }

    return new WorkingMemoryPullQuery<string>(tokens[1], tokens[2], tokens[3], 
					      0,
					      new string(tokens[4]), 
					      AbstractWorkingMemoryPullQuery::OVERWRITE_COUNT);
  }

  WorkingMemoryPullQuery< vector<string> > * 
  SubarchitectureWorkingMemoryProtocol::extractIDArray(const string &_query) {

    vector<string> tokens;

    tokenizeString(_query, tokens, "\\");
  
    if(tokens.size() < 5) {
      cerr<<"SubarchitectureWorkingMemoryProtocol::extractIDArray incorrect number of tokens in query"<<endl;
      return NULL;
    }
  
    vector<string> *pIDs = new vector<string>();
    for(unsigned int i = 4; i < tokens.size(); i++) {
      pIDs->push_back(tokens[i]);
    }
  
    return new WorkingMemoryPullQuery< vector<string> >(tokens[1], tokens[2], tokens[3], 
							0,
							pIDs, 
							AbstractWorkingMemoryPullQuery::ID_ARRAY);
  }

  WorkingMemoryPullQuery<string> * 
  SubarchitectureWorkingMemoryProtocol::extractType(const string &_query) {
    vector<string> tokens;
  
    tokenizeString(_query, tokens, "\\");
  
    if(6 != tokens.size()) {
      cerr<<"SubarchitectureWorkingMemoryProtocol::extractType incorrect number of tokens in query"<<endl;
      return NULL;
    }

    return new WorkingMemoryPullQuery<string>(tokens[1], tokens[2], tokens[3], 
					      atoi(tokens[4].c_str()),
					      new string(tokens[5]), 
					      AbstractWorkingMemoryPullQuery::TYPE);
  }


  /**
   * Stolen from http://www.linuxselfhelp.com/HOWTO/C++Programming-HOWTO-7.html
   */
  void 
  SubarchitectureWorkingMemoryProtocol::tokenizeString(const string & _str,
						       vector < string > &_tokens,
						       const string & _delimiters) {
    // Skip delimiters at beginning.
    string::size_type lastPos = _str.find_first_not_of(_delimiters, 0);
    // Find first "non-delimiter".
    string::size_type pos = _str.find_first_of(_delimiters, lastPos);
  
    while(string::npos != pos || string::npos != lastPos) {
      // Found a token, add it to the vector.
      _tokens.push_back(_str.substr(lastPos, pos - lastPos));
      // Skip delimiters.  Note the "not_of"
      lastPos = _str.find_first_not_of(_delimiters, pos);
      // Find next "non-delimiter"
      pos = _str.find_first_of(_delimiters, lastPos);
    }
  }


  string SubarchitectureWorkingMemoryProtocol::createQuery(AbstractWorkingMemoryPullQuery * _pQuery) {

    if(_pQuery->getType() == AbstractWorkingMemoryPullQuery::ID) {
   
      WorkingMemoryPullQuery<string> * pQuery = 
	dynamic_cast<WorkingMemoryPullQuery<string> *>(_pQuery);
      if(pQuery) {
	return createIDQuery(pQuery->getSourceID(),pQuery->getSourceSA(),pQuery->getSubarchitectureID(),*pQuery->getQueryObject());
      }
      else {
	cerr<<"ERROR: casting query to ID"<<endl;
	return "";
      }

    }
    else if(_pQuery->getType() == AbstractWorkingMemoryPullQuery::EXISTS) {
   
      WorkingMemoryPullQuery<string> * pQuery = 
	dynamic_cast<WorkingMemoryPullQuery<string> *>(_pQuery);
      if(pQuery) {
	return createExistsQuery(pQuery->getSourceID(),pQuery->getSourceSA(),pQuery->getSubarchitectureID(),*pQuery->getQueryObject());
      }
      else {
	cerr<<"ERROR: casting query to EXISTS"<<endl;
	return "";
      }

    }
    else if(_pQuery->getType() == AbstractWorkingMemoryPullQuery::OVERWRITE_COUNT) {
   
      WorkingMemoryPullQuery<string> * pQuery = 
	dynamic_cast<WorkingMemoryPullQuery<string> *>(_pQuery);
      if(pQuery) {
	return createOverwriteCountQuery(pQuery->getSourceID(),pQuery->getSourceSA(),pQuery->getSubarchitectureID(),*pQuery->getQueryObject());
      }
      else {
	cerr<<"ERROR: casting query to OVERWRITE_COUNT"<<endl;
	return "";
      }

    }
    else if(_pQuery->getType() == AbstractWorkingMemoryPullQuery::ID_ARRAY) {
      //cerr<<"ID Array query"<<endl;
      WorkingMemoryPullQuery< vector<string> > * pQuery = 
	dynamic_cast<WorkingMemoryPullQuery< vector<string> > *>(_pQuery);
      if(pQuery) {
	return createIDArrayQuery(pQuery->getSourceID(),pQuery->getSourceSA(),pQuery->getSubarchitectureID(),*pQuery->getQueryObject());
      }
      else {
	cerr<<"ERROR: casting query to ID array"<<endl;
	return "";
      }

    }
    else if(_pQuery->getType() == AbstractWorkingMemoryPullQuery::TYPE) {
      //cerr<<"Type query"<<endl;
      WorkingMemoryPullQuery<string> * pQuery = 
	dynamic_cast<WorkingMemoryPullQuery<string> *>(_pQuery);
      if(pQuery) {
	return createTypeQuery(pQuery->getSourceID(),pQuery->getSourceSA(),pQuery->getSubarchitectureID(),*pQuery->getQueryObject(), pQuery->getCount());
      }
      else {
	cerr<<"ERROR: casting query to type"<<endl;
	return "";
      }

    }
    else if(_pQuery->getType() == AbstractWorkingMemoryPullQuery::OTHER) {
      //cerr<<"Other query"<<endl;
      WorkingMemoryPullQuery<string> * pQuery = 
	dynamic_cast<WorkingMemoryPullQuery<string> *>(_pQuery);
      if(pQuery) {
	return string(*pQuery->getQueryObject());
      }
      else {
	cerr<<"ERROR: casting query to other"<<endl;
	return "";
      }

    }
    else {
      cerr<<"Unknown query type"<<endl;
      return "";
    }

  }

} //namespace cast

File: binding/development/tools/cast/src/java/cast/architecture/subarchitecture/SubarchitectureWorkingMemoryProtocol.java
/*
 * CAST - The CoSy Architecture Schema Toolkit
 *
 * Copyright (C) 2006-2007 Nick Hawes
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA
 *
 */

/**
 * 
 */
package cast.architecture.subarchitecture;

import java.util.NoSuchElementException;
import java.util.StringTokenizer;

import cast.architecture.subarchitecture.SubarchitectureWorkingMemoryProtocol.WorkingMemoryPullQuery.WorkingMemoryQueryType;
import cast.cdl.*;

/**
 * Manages the language used for pull queries from the
 * SubarchitectureWorkingMemory. We may need a better long term solution
 * though! The key behaviours are as follows - pull from an indicated
 * working memory
 * 
 * @author nah
 */
public class SubarchitectureWorkingMemoryProtocol {

    public static class WorkingMemoryPullQuery {

        public enum WorkingMemoryQueryType {
            TYPE, ID, EXISTS, ID_ARRAY, OVERWRITE_COUNT, OTHER
        };

        // ID of the sub-architecture
        private String m_saID;
        // max number of results requested
        private int m_count;
        // query object
        private Object m_queryObject;
        // and the type
        private WorkingMemoryQueryType m_type;
        private String m_srcID;
        private String m_srcSA;

        /**
         * @return the srcID
         */
        public String getSourceID() {
            return m_srcID;
        }

        /**
         * @return the srcSA
         */
        public String getSourceSA() {
            return m_srcSA;
        }

        /**
         * @param __said
         * @param __count
         * @param _object
         */
        public WorkingMemoryPullQuery(String _srcID, String _srcSA,
                String __said, int __count, Object _object,
                WorkingMemoryQueryType _type) {
            m_srcID = _srcID;
            m_srcSA = _srcSA;

            m_saID = __said;
            m_count = __count;
            m_queryObject = _object;
            m_type = _type;
        }

        public WorkingMemoryQueryType getType() {
            return m_type;
        }

        public int getCount() {
            return m_count;
        }

        public Object getQueryObject() {
            return m_queryObject;
        }

        public String getSubarchitectureID() {
            return m_saID;
        }

    }

    protected static final String ID_PREFIX = ID_QUERY_PREFIX.value;
    protected static final String EXISTS_PREFIX =
            EXISTS_QUERY_PREFIX.value;
    
    protected static final String OVERWRITE_COUNT_PREFIX =
        OVERWRITE_COUNT_QUERY_PREFIX.value;

    
    protected static final String ID_ARRAY_PREFIX =
            ID_ARRAY_QUERY_PREFIX.value;
    protected static final String TYPE_PREFIX = TYPE_QUERY_PREFIX.value;

    /**
     * Query working memory by particular index.
     * 
     * @param _i
     * @return
     */
    public static String createIDQuery(String _srcID,
                                       String _srcSA,
                                       String _subarch,
                                       String _id) {
        return ID_PREFIX + "\\" + _srcID + "\\" + _srcSA + "\\"
            + _subarch + "\\" + _id;
    }

    public static String createExistsQuery(String _srcID,
                                           String _srcSA,
                                           String _subarch,
                                           String _id) {
        return EXISTS_PREFIX + "\\" + _srcID + "\\" + _srcSA + "\\"
            + _subarch + "\\" + _id;
    }

    public static String createOverwriteCountQuery(String _srcID,
                                           String _srcSA,
                                           String _subarch,
                                           String _id) {
        return OVERWRITE_COUNT_PREFIX + "\\" + _srcID + "\\" + _srcSA + "\\"
            + _subarch + "\\" + _id;
    }
    
    /**
     * Query working memory by a collection of indices.
     * 
     * @param _is
     * @return
     */
    public static String createIDArrayQuery(String _srcID,
                                            String _srcSA,
                                            String _subarch,
                                            String[] _ids) {
        String s =
                ID_ARRAY_PREFIX + "\\" + _srcID + "\\" + _srcSA + "\\"
                    + _subarch + "\\";
        for (int i = 0; i < _ids.length; i++) {
            s += _ids[i] + "\\";
        }
        return s;
    }

    /**
     * Query working memory by ontological type. Restrict number of
     * results. Starts from most recent.
     * 
     * @param _type
     * @param _count
     * @return
     */
    public static String createTypeQuery(String _srcID,
                                         String _srcSA,
                                         String _subarch,
                                         String _type,
                                         int _count) {
        return TYPE_PREFIX + "\\" + _srcID + "\\" + _srcSA + "\\"
            + _subarch + "\\" + _count + "\\" + _type;
    }

    public static WorkingMemoryPullQuery parseQuery(String _query)
            throws SubarchitectureProcessException {

        // System.out
        // .println("SubarchitectureWorkingMemoryProtocol.parseQuery():
        // " + _query);

        if (_query.startsWith(ID_PREFIX)) {
            return extractID(_query);
        }
        else if (_query.startsWith(EXISTS_PREFIX)) {
            return extractExists(_query);
        }
        else if (_query.startsWith(OVERWRITE_COUNT_PREFIX)) {
            return extractOverwriteCount(_query);
        }
        else if (_query.startsWith(ID_ARRAY_PREFIX)) {
            return extractIDArray(_query);
        }
        else if (_query.startsWith(TYPE_PREFIX)) {
            return extractType(_query);
        }
        else {
            return extractUnknown(_query);
        }

    }

    protected static WorkingMemoryPullQuery extractUnknown(String _query)
            throws SubarchitectureProcessException {

        // System.out
        // .println("SubarchitectureWorkingMemoryProtocol.extractID()");
        // System.out.println(_query);

        try {
            StringTokenizer tokenizer =
                    new StringTokenizer(_query, "\\");
            // skip prefix token
            tokenizer.nextToken();

            // assume these 3 are ok... the rest can be unknown
            String srcID = tokenizer.nextToken();
            String srcSA = tokenizer.nextToken();
            String targetSA = tokenizer.nextToken();

            return new WorkingMemoryPullQuery(srcID, srcSA, targetSA,
                0, _query, WorkingMemoryQueryType.OTHER);

        }
        catch (NoSuchElementException e) {
            throw new SubarchitectureProcessException(
                "Unknown memory protocol for query: " + _query, e);
        }
    }

    protected static WorkingMemoryPullQuery extractID(String _query) {

        // System.out
        // .println("SubarchitectureWorkingMemoryProtocol.extractID()");
        // System.out.println(_query);

        StringTokenizer tokenizer = new StringTokenizer(_query, "\\");
        // skip prefix token
        tokenizer.nextToken();

        String srcID = tokenizer.nextToken();
        String srcSA = tokenizer.nextToken();
        String subarch = tokenizer.nextToken();
        String intString = tokenizer.nextToken();

        return new WorkingMemoryPullQuery(srcID, srcSA, subarch, 0,
            intString, WorkingMemoryQueryType.ID);

    }

    protected static WorkingMemoryPullQuery extractExists(String _query) {

        // System.out
        // .println("SubarchitectureWorkingMemoryProtocol.extractID()");
        // System.out.println(_query);

        StringTokenizer tokenizer = new StringTokenizer(_query, "\\");
        // skip prefix token
        tokenizer.nextToken();

        String srcID = tokenizer.nextToken();
        String srcSA = tokenizer.nextToken();
        String subarch = tokenizer.nextToken();
        String intString = tokenizer.nextToken();

        return new WorkingMemoryPullQuery(srcID, srcSA, subarch, 0,
            intString, WorkingMemoryQueryType.EXISTS);

    }

    protected static WorkingMemoryPullQuery extractOverwriteCount(String _query) {

        // System.out
        // .println("SubarchitectureWorkingMemoryProtocol.extractID()");
        // System.out.println(_query);

        StringTokenizer tokenizer = new StringTokenizer(_query, "\\");
        // skip prefix token
        tokenizer.nextToken();

        String srcID = tokenizer.nextToken();
        String srcSA = tokenizer.nextToken();
        String subarch = tokenizer.nextToken();
        String intString = tokenizer.nextToken();

        return new WorkingMemoryPullQuery(srcID, srcSA, subarch, 0,
            intString, WorkingMemoryQueryType.OVERWRITE_COUNT);

    }
    
    protected static WorkingMemoryPullQuery extractIDArray(String _query) {
        StringTokenizer tokenizer = new StringTokenizer(_query, "\\");
        // skip prefix token
        tokenizer.nextToken();

        String srcID = tokenizer.nextToken();
        String srcSA = tokenizer.nextToken();
        String subarch = tokenizer.nextToken();

        String ia[] = new String[tokenizer.countTokens()];
        int i = 0;

        while (tokenizer.hasMoreTokens()) {
            ia[i++] = tokenizer.nextToken();
        }

        return new WorkingMemoryPullQuery(srcID, srcSA, subarch, 0, ia,
            WorkingMemoryQueryType.ID_ARRAY);

    }

    protected static WorkingMemoryPullQuery extractType(String _query) {

        // STRING_PREFIX + "\\" + _subarch + "\\" + _count + "\\" + _s;

        StringTokenizer tokenizer = new StringTokenizer(_query, "\\");

        // skip prefix token
        tokenizer.nextToken();

        String srcID = tokenizer.nextToken();
        String srcSA = tokenizer.nextToken();

        String subarch = tokenizer.nextToken();

        String count = tokenizer.nextToken();

        String typeString = tokenizer.nextToken();

        return new WorkingMemoryPullQuery(srcID, srcSA, subarch,
            Integer.parseInt(count), typeString,
            WorkingMemoryQueryType.TYPE);
    }

    public static void main(String[] args) {
    // String query = memoryQuery(999);
    // System.out.println(query);
    // int i = extractInt(query);
    // System.out.println(i);

    // String query = memoryQuery(new int []{999,1, 22, 345});
    // System.out.println(query);
    // int ia[] = extractIntArray(query);
    // for (int i = 0; i < ia.length; i++) {
    // System.out.println(ia[i]);
    // }

    // String query = memoryQuery("blah");
    // System.out.println(query);
    // String s = extractString(query);
    // System.out.println(s);

    }

    /**
     * @param _queryObject
     * @return
     * @throws SubarchitectureProcessException
     * @throws SubarchitectureProcessException
     */
    public static String createQuery(WorkingMemoryPullQuery _queryObject)
            throws SubarchitectureProcessException {
        if (_queryObject.getType() == WorkingMemoryQueryType.ID) {
            return createIDQuery(_queryObject.getSourceID(),
                _queryObject.getSourceSA(), _queryObject
                    .getSubarchitectureID(), (String) _queryObject
                    .getQueryObject());
        }
        else if (_queryObject.getType() == WorkingMemoryQueryType.EXISTS) {
            return createExistsQuery(_queryObject.getSourceID(),
                _queryObject.getSourceSA(), _queryObject
                    .getSubarchitectureID(), (String) _queryObject
                    .getQueryObject());
        }
        else if (_queryObject.getType() == WorkingMemoryQueryType.OVERWRITE_COUNT) {
            return createOverwriteCountQuery(_queryObject.getSourceID(),
                _queryObject.getSourceSA(), _queryObject
                    .getSubarchitectureID(), (String) _queryObject
                    .getQueryObject());
        }
        else if (_queryObject.getType() == WorkingMemoryQueryType.ID_ARRAY) {
            return createIDArrayQuery(_queryObject.getSourceID(),
                _queryObject.getSourceSA(), _queryObject
                    .getSubarchitectureID(), (String[]) _queryObject
                    .getQueryObject());
        }
        else if (_queryObject.getType() == WorkingMemoryQueryType.TYPE) {
            return createTypeQuery(_queryObject.getSourceID(),
                _queryObject.getSourceSA(), _queryObject
                    .getSubarchitectureID(), (String) _queryObject
                    .getQueryObject(), _queryObject.getCount());
        }
        else if (_queryObject.getType() == WorkingMemoryQueryType.OTHER) {
            return (String) _queryObject.getQueryObject();
        }

        throw new SubarchitectureProcessException(
            "Unknown memory protocol for query: "
                + _queryObject.getType());
    }

}
